### PR TITLE
feat(dataset): importer

### PR DESCRIPTION
**Pull request by reynld**
Allows users to import datasets

### Accepted params
`--clusterId` Cluster to run import on
`--machineType` Machine to run import job on
`--datasetId` Dataset ID
`--s3Url` URL for s3 URLs
`--httpUrl` URL for s3 URLs
`--s3AccessKey` s3 Access Key
`--s3SecretKey` s3 Secret Key
`--httpAuth` HTTP Basic auth username:password

#### HTTPS
```
python3 ./gradient datasets import \
--clusterId clcjo828k \
--machineType c5.xlarge \
--datasetId dsr4bvtdcsefhgq \
--httpUrl google.com 
```
#### S3
```
python3 ./gradient datasets import \
--clusterId clcjo828k \
--machineType c5.xlarge \
--datasetId dsr4bvtdcsefhgq \
--s3Url s3://talmnist/tryagain2.csv \
--s3AccessKey ACCESS_KEY \
--s3SecretKey SECRET
```
### QA TEST PLAN
- Set PAPERSPACE_CONFIG_HOST environment variable to point to staging
`export PAPERSPACE_CONFIG_HOST=https://staging-api.paperspace.io`
- Create dataset
Staging Google Saml Prod storage provider id: `sp694axp9oevzyj`
`python3 ./gradient datasets create --name <NAME> --storageProviderId <STORAGE PROVIDER ID>`
- Run import job using the examples above

### CODE REVIEW

File: gradient/cli/datasets.py [hunk 1]
```diff
@@ -143,6 +148,67 @@ def update_dataset(
...
+    "--datasetId",
+    "dataset_id",
+    help="Dataset ID",
+    cls=common.GradientOption,
+    required=True,
+)
+@click.option(
+    "--url",
+    "dataset_url",
+    help="Dataset URL",
```
Comment (paperspace-philip) [inline gradient/cli/datasets.py:176]: It's just url, it's not a dataset until the job runs

File: gradient/cli/datasets.py [hunk 2]
```diff
@@ -143,6 +148,67 @@ def update_dataset(
...
+    cls=common.GradientOption,
+    required=True,
+)
+@click.option(
+    "--type",
+    "url_type",
+    help="Dataset Type s3|git|https",
```
Comment (paperspace-philip) [inline gradient/cli/datasets.py:183]: It would be nicer to accept URIs and then parse the protocol from that

File: gradient/cli/datasets.py [hunk 3]
```diff
@@ -143,6 +148,67 @@ def update_dataset(
...
+    cls=common.GradientOption,
+    required=True,
+)
+@click.option(
+    "--httpAuth",
+    "http_auth",
+    help="Dataset HttpAuth username:password",
```
Comment (paperspace-philip) [inline gradient/cli/datasets.py:190]: HTTPAuth, not a dataset yet

File: gradient/commands/datasets.py [hunk 4]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+
+        response = client.ephemral(key, value, expires_in)
+        return response
+    @classmethod
+    def get_command_string(self, url):
+        return "go-getter {} /data/output".format(url)
+
+    @classmethod
+    def get_command(self, dataset_url, url_type, http_auth, access_key, secret_key):
+        if url_type == 'git' or url_type == 's3':
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:704]: you can determine the protocol from the url

File: gradient/commands/datasets.py [hunk 5]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+            return self.get_command_string(http_auth_url)
+
+        return ""
+    @classmethod
+    def get_env_vars(self, url_type, http_auth, access_key, secret_key):
+        if url_type == 'git':
+            return ""
+
+        if url_type == 's3':
+            access_key_secret = self.create_secret('AWS_ACCESS_KEY_ID', access_key)
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:721]: If public S3 URL, no access key or secret is required

File: gradient/commands/datasets.py [hunk 6]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+            secret_key_secret = self.create_secret('AWS_ACCESS_KEY_SECRET', secret_key)
+
+            access_key_value = "secret:ephemeral:{}".format(access_key_secret["AWS_ACCESS_KEY_ID"])
+            secret_key_value = "secret:ephemeral:{}".format(secret_key_secret["AWS_ACCESS_KEY_SECRET"])
+
+            return "{\"AWS_ACCESS_KEY_ID\":\"%s\", \"AWS_ACCESS_KEY_SECRET\":\"%s\"}" % (access_key_value, secret_key_value)
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:727]: Return a python dict and encode to json later

File: gradient/commands/datasets.py [hunk 7]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+
+        if url_type == 'https':
+            http_auth_secret = self.create_secret('HTTP_AUTH', http_auth)
+            return "{\"HTTP_AUTH\":\"%s\"}" % http_auth_secret
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:731]: can't you return a python dict and encode later?

File: gradient/commands/datasets.py [hunk 8]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+        client = api_sdk.clients.JobsClient(
+            api_key="ae2ea576ef51c046295bec0d004c67",
+            ps_client_name="etest",
+        )
+        client.create(**workflow)
+
+    def execute(self, cluster_id, machine_type, dataset_id, dataset_url, url_type, http_auth, access_key, secret_key):
+        workflow = {
+            "cluster_id": cluster_id,
+            "container": "paperspace/dataset-importer:latest",
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:745]: this could be better as a constant at the top

File: gradient/commands/datasets.py [hunk 9]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+            "machine_type": machine_type,
+            "project": "Job Builder",
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:747]: This could be better as a constant up top

File: gradient/commands/datasets.py [hunk 10]
```diff
@@ -676,3 +678,83 @@ def update_status():
...
+            "datasets": [{ "id": dataset_id, "name": "output", "output": True }],
+            "project_id": None
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:749]: is this needed?

File: gradient/commands/datasets.py (hunk 6 above)
Comment (reynld) [reply gradient/commands/datasets.py:727]: env variables needs to be a string

File: gradient/commands/datasets.py (hunk 7 above)
Comment (reynld) [reply gradient/commands/datasets.py:731]: same as above

File: gradient/commands/datasets.py (hunk 10 above)
Comment (reynld) [reply gradient/commands/datasets.py:749]: yes

File: gradient/cli/datasets.py [hunk 11]
```diff
@@ -143,6 +148,60 @@ def update_dataset(
...
+@click.option(
+    "--datasetId",
+    "dataset_id",
+    help="Dataset ID",
+    cls=common.GradientOption,
+    required=True,
+)
+@click.option(
+    "--url",
+    "dataset_url",
```
[the remark this answers is not in the document]
Comment (reynld) [reply gradient/cli/datasets.py:175]: its the url for all s3, https, git

File: gradient/commands/datasets.py [hunk 12]
```diff
@@ -676,3 +680,91 @@ def update_status():
...
+
+            access_key_value = "secret:ephemeral:{}".format(access_key_secret["AWS_ACCESS_KEY_ID"])
+            secret_key_value = "secret:ephemeral:{}".format(secret_key_secret["AWS_ACCESS_KEY_SECRET"])
+
+            return {
+                "AWS_ACCESS_KEY_ID": access_key_value,
+                "AWS_ACCESS_KEY_SECRET": secret_key_value
+            }
+
+        if url.scheme == 'https':
```
[the remark this answers is not in the document]
Comment (reynld) [reply gradient/commands/datasets.py:729]: added :)

File: gradient/commands/datasets.py [hunk 13]
```diff
@@ -676,3 +680,91 @@ def update_status():
...
+
+    def get_command_string(self, url):
+        return "go-getter {} /data/output".format(url)
+
+
+    def get_command(self, url, http_auth, access_key, secret_key):
+        if url.scheme == 'git' or url.scheme == 's3':
+            return self.get_command_string(url.geturl())
+        
+        if url.scheme == 'https':
```
[the remark this answers is not in the document]
Comment (reynld) [reply gradient/commands/datasets.py:705]: added :)

File: gradient/commands/datasets.py [hunk 14]
```diff
@@ -22,6 +24,13 @@
 from gradient.exceptions import ApplicationError
 
 S3_XMLNS = 'http://s3.amazonaws.com/doc/2006-03-01/'
+DATASET_IMPORTER_IMAGE = "paperspace/dataset-importer:latest"
+PROJECT_NAME = "Job Builder"
+SUPPORTED_URL = "['s3', 'git', 'https', 'http']"
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:29]: Why is this a string?

File: gradient/commands/datasets.py [hunk 15]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+        response = client.ephemeral(key, value, expires_in)
+        return response
+
+
+    def get_command_string(self, url):
+        return "%s %s /data/output" % (IMPORTER_COMMAND, url)
+
+
+    def get_command(self, url, http_auth):
+
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:707]: nit: why is this here?

File: gradient/commands/datasets.py [hunk 16]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+        return response
+
+
+    def get_command_string(self, url):
+        return "%s %s /data/output" % (IMPORTER_COMMAND, url)
+
+
+    def get_command(self, url, http_auth):
+        
+        if url.scheme == 'git' or url.scheme == 's3':
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:708]: isn't it easier to write: command_url = url.geturl()

then do conditionals to change it and return at the end?

File: gradient/commands/datasets.py [hunk 17]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+        if url.scheme == 'https' or url.scheme == 'http':
+            if not http_auth:
+                return self.get_command_string(url.geturl())
+            http_auth_url = "https://${{HTTP_AUTH}}@%s" % url.path
+            return self.get_command_string(http_auth_url)
+
+        return ""
+    
+    def get_env_vars(self, url, secrets):
+        # http_auth, access_key, secret_key
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:720]: ?

File: gradient/commands/datasets.py [hunk 18]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:721]: nit: space

File: gradient/commands/datasets.py [hunk 19]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+        if url.scheme == 'git':
+            return ""
+
+        if url.scheme == 's3' and secrets[S3_ACCESS_KEY] and secrets[S3_SECRET_KEY]:
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:725]: if url.scheme == 's3' and S3_ACCESS_KEY in secrets and S3_SECRET_KEY in secrets:

File: gradient/commands/datasets.py [hunk 20]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+
+            access_key_value = "secret:ephemeral:%s" % access_key_secret[S3_ACCESS_KEY]
+            secret_key_value = "secret:ephemeral:%s" % secret_key_secret[S3_SECRET_KEY]
+
+            return {
+                S3_ACCESS_KEY: access_key_value,
+                S3_SECRET_KEY: secret_key_value
+            }
+
+        if url.scheme == 'https' and secrets[HTTP_SECRET]:
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:737]: HTTP_SECRET in secrets

File: gradient/commands/datasets.py [hunk 21]
```diff
@@ -676,3 +685,94 @@ def update_status():
...
+            "machine_type": machine_type,
+            "project": PROJECT_NAME,
+            "datasets": [{ "id": dataset_id, "name": "output", "output": True }],
+            "project_id": None
+        }
+
+        url = urlparse(dataset_url)
+
+        if url.scheme not in SUPPORTED_URL:
+            self.logger.log('Invalid URL format supported [ git | s3 | https | http ]: {}'.format(dataset_url))
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:766]: you can use the supported type in the string using join

File: gradient/cli/datasets.py [hunk 22]
```diff
@@ -143,6 +148,60 @@ def update_dataset(
         description=description,
     )
 
+@datasets.command("import", help="Import dataset")
+@click.option(
+    "--clusterId",
+    "cluster_id",
+    help="Cluster id",
```
Comment (paperspace-philip) [inline gradient/cli/datasets.py:155]: nit: lower case id , but all caps below

File: gradient/commands/datasets.py (hunk 21 above)
Comment (reynld) [reply gradient/commands/datasets.py:766]: 👍

File: gradient/commands/datasets.py [hunk 23]
```diff
@@ -676,3 +686,83 @@ def update_status():
...
+        if url.scheme not in SUPPORTED_URL:
+            self.logger.log('Invalid URL format supported [{}] Format:{} URL:{}'.format(','.join(SUPPORTED_URL), url.scheme, dataset_url))
+            return
+
+
+        command = self.get_command(url, http_auth)
+        if command:
+            workflow["command"] = command
+
+        env_vars = self.get_env_vars(url, {HTTP_SECRET: http_auth, S3_ACCESS_KEY: access_key, S3_SECRET_KEY: secret_key})
```
Comment (paperspace-philip) [inline gradient/commands/datasets.py:764]: nit: add space between { } like above

**Review by paperspace-philip — APPROVED** [review]

File: pyproject.toml [hunk 24]
```diff
@@ -0,0 +1,3 @@
+[build-system]
```
Comment (paperspace-philip) [inline pyproject.toml:1]: what is this file?

**Review by paperspace-philip — APPROVED** [review]
